### ISSUE TITLE
SetMaxTerms(-1) yields wrong results

### ISSUE THREAD

**Issue by 10110111** [ovidiopr/scattnlay#44]
I'm not sure what motivation for #23 was, but the following gives a test case that shows that current automatic determination of the number of required terms works incorrectly.

Consider the following calculation:
```
$ ./fieldnlay-mp -l 1 2 1.3 0 -p 1e-3 1e-3 1 1e-3 1e-3 1 -200 200 2000 | cut -d, -f3,4 > /tmp/field.csv
```
This gives us _ℜ{Eₓ}_ vs _z_ table. The result looks as follows:

![Screenshot - 281121 - 14:10:27](https://user-images.githubusercontent.com/6376882/143765391-c06e9299-931d-48f0-a085-5f0331499e8a.png)

This doesn't make much sense: we should see incident wave with a bit of interference due to the scattered wave, rather than the widely-changing envelope we actually got. Let's now explicitly set 100 terms in `nearfield.cc`:
```diff
diff --git a/src/nearfield.cc b/src/nearfield.cc
index e842115..a914248 100644
--- a/src/nearfield.cc
+++ b/src/nearfield.cc
@@ -235,7 +235,7 @@ int main(int argc, char *argv[]) {
       }
     }
 
-    nmie::nField(L, -1, x, m, -1, nmie::Modes::kAll, nmie::Modes::kAll, total_points, Xp, Yp, Zp, E, H);
+    nmie::nField(L, -1, x, m, 100, nmie::Modes::kAll, nmie::Modes::kAll, total_points, Xp, Yp, Zp, E, H);
 
     if (has_comment)
       printf("%6s\n", comment.c_str());
```
Now the results look much more sensible. The incident wave becomes visible much farther from the sphere, although it still goes wild at some distance:

![Screenshot - 281121 - 14:16:43](https://user-images.githubusercontent.com/6376882/143765566-23f04df7-8284-4c89-89ed-64bd965a4e59.png)

Changing the maximum to 250 gives the good expected result:

![Screenshot - 281121 - 14:25:50](https://user-images.githubusercontent.com/6376882/143765792-1a8f4377-3096-4f9e-8fa2-f4526e76843f.png)

**Comment by kostyfisik** [comment]
Are you using the latest master branch from Github? It should be printing a warning message if the convergence is poor. Note, Le Ru nmax estimate is correct for near field inside a bulk sphere an space about one radius away from it.

**Comment by 10110111** [comment]
Yes, I'm using the latest master. As for the warning, I do get one, but I thought it was about the single failing point discussed in #42.

Is there any available way to estimate nmax for larger distance from the sphere?

**Comment by kostyfisik** [comment]
No, Le Ru seems to be the latest update on this.  In general, near-field evaluation is a poorly investigated aspect of Mie theory. The next problem I would like to adress here, is to develop a more stable algorithm for near-field evaluation inside the sphere. As for the near-field outside the sphere it is way more stable than, e.g. in Matlab implementations, or some published results (wich are simply wrong). In the attached figre you can find Stratify result on the top left (note halo-type problem away from the sphere due to use of Wiscombe cutoff), our Scattnlay result on the right (with Le Ru cutoff), and published result in the background.
![near-field](https://user-images.githubusercontent.com/5239997/143783020-36db040c-8b98-4250-9c7f-b9326b4df18f.jpg)

**Comment by 10110111** [comment]
I think the problem here is that incident field is calculated from the spherical partial wave expansion, using the termination criterion that is designed for scattered field properties, but is inadequate for incident field. (Le Ru doesn't even discuss incident field.)

It may be more efficient (i.e. less demanding on number of terms) to calculate the incident field as an actual exp(ikz) function, instead of summing its expansion in spherical waves along with the scattered near field.

**Comment by kostyfisik** [comment]
@10110111 Good idea! Could you please plot separately the incident and scattered fields for some set of parameters, when it really helps? You can use https://github.com/ovidiopr/scattnlay/blob/master/src/nmie-nearfield.hpp#L402 as a blue-print for your code. Note, that we use the same summation cycle both for incident and scattered field incide and outside the particle https://github.com/ovidiopr/scattnlay/blob/master/src/nmie-nearfield.hpp#L330-L334 , we just set c and d expansion coefficients to be one https://github.com/ovidiopr/scattnlay/blob/master/src/nmie-nearfield.hpp#L107-L108

**Comment by 10110111** [comment]
I've actually now implemented my idea using your pointers (see the patch here: https://github.com/10110111/scattnlay/commit/6c8406409c44fed1d8952911fba468b132724a50), and the results look the same as the last screenshot in https://github.com/ovidiopr/scattnlay/issues/44#issue-1065307022 — i.e. ideally:

![Screenshot - 201221 - 15:17:34](https://user-images.githubusercontent.com/6376882/146766056-ead562f7-4bd9-4279-81bc-f06901887568.png)

In case you still want to see the scattered field only, here it is (the parameters are the same as in the OP):

![Screenshot - 201221 - 15:07:57](https://user-images.githubusercontent.com/6376882/146765066-76ce6b7c-fa4a-47bc-9936-a86a282491d9.png)

And here's the current `master` output for incident-only field (I just subtracted the scattered field from the full output):

![Screenshot - 201221 - 15:16:45](https://user-images.githubusercontent.com/6376882/146765965-f3bcf813-89f3-4958-b138-31e10e05e40f.png)

**Comment by kostyfisik** [comment]
Great! Looks very promising. Still need some verification for a general case (there is still a possibility e.g. messing a sign in some direction); however, it looks like a way to go... I would like to finish with adding near- and far-field computations to the webapp and will work on this next...
